src/metatrain/utils/additive/composition.py

```diff
@@ -155,17 +152,39 @@ def train_model(
                     )
 
                 composition_features = torch.zeros(
-                    (total_num_structures, len(self.atomic_types)), dtype=dtype
+                    (total_num_structures, len(self.atomic_types)),
+                    dtype=dtype,
+                    device=device,
                 )
-                structure_index = 0
+                system_index = 0
+                targets_list = []
+
                 for dataset in datasets_with_target:
                     for sample in dataset:
-                        structure = sample["system"]
+                        systems = [sample["system"]]
+                        targets = {target_key: sample[target_key]}
+                        systems, targets = systems_and_targets_to_device(
+                            systems, targets, device
+                        )
+                        for additive_model in additive_models:
```

Comment:
I mean, I would move this for-loop out of the composition model class, it feels like this functionality goes out of the scope of the CompositionModel